sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala

```diff
@@ -581,4 +581,40 @@ class UDFSuite extends QueryTest with SharedSparkSession {
       .toDF("col1", "col2")
     checkAnswer(df.select(myUdf(Column("col1"), Column("col2"))), Row(2020) :: Nil)
   }
+
+  test("case class as element type of Seq/Array") {
```

Comment:
some option tests would be good too. 

option in a simple seq:
`Seq(Seq(Some(1), None)).toDF.withColumn("value", udf{ s: Seq[Option[Int]] => s.map(_.map(_ + 1)) }.apply(col("value")) )`

option for function argument:
`Seq(None, Some(1), None).toDF.withColumn("value", udf{ o: Option[Int] => o.map(_ + 1) }.apply(col("value")))`

note that the top level option to express nullability is a very common use case in particular and supported by encoders. the equivalent in Dataset is:
`Seq(None, Some(1), None).toDS.map{ o: Option[Int] => o.map(_ + 1) }`

Reply:
There's implementation difference between `udf` and `Dataset.map`. So for the second case you're mentioned, it only work in  `Dataset.map` but fail in `udf`.

Reply:
@koertkuipers I've updated to support the second case.